magma/circuit.py

```diff
@@ -321,6 +319,8 @@ def __new__(metacls, name, bases, dct):
             # Override staged context with '_context_' from namespace if
             # available.
             cls._context_ = dct.get("_context_", context)
+            cls._context_.place_instances(cls)
```

Comment:
is there a reason to split these? It seems best to have everything encapsulated in finalize.

Reply:
In the old-style syntax, we need to call setup_interface before running the definition, then we place the instances afterwards.  In the new-style syntax, we want to call setup_interface after placing instances so we can add the automatically lifted clock if necessary.  For the old-style syntax, we need to call _setup_interface before the definition because otherwise the definition won't have the write ports to wire up.  For the new-style syntax, we resolve this with the staged placer logic which handles everything after definition.

Reply:
Makes sense. Hoping we can completely clean out old-style stuff in v2, so it might simplify it then.